### PR TITLE
Split up project in library and binary

### LINKED ISSUE #19: Cleanup error handling
The binary should not panic on errors but simply write information to `stderr` and return an exit code different than 0.

### LINKED ISSUE #22: Split up sources and create library
> Also a more general note: I saw that you have a `main.rs` and then a `gccrs` module.
> Usually you just make a lib and add one small `main.rs` to keep this split. You could do this, by moving everything from `src/gccrs/*` to `src/` and rename `mod.rs` to `lib.rs`

_Originally posted by @flip1995 in https://github.com/Rust-GCC/cargo-gccrs/issues/21#issuecomment-866028039_

### PR DESCRIPTION
**Pull request by CohenArthur**
Closes #19 
Closes #22 
In reference to [this comment](https://github.com/Rust-GCC/cargo-gccrs/pull/15#pullrequestreview-688108719)

@flip1995 I am unsure about the implementation as I usually just panic when handling errors in the main haha. I believe this to be cleaner, as it displays useful information and exits using `std::process::exit()` rather than panicking.

### CODE REVIEW

**Comment by flip1995** [comment]
Did you look into if using [`anyhow`](https://github.com/dtolnay/anyhow) and/or [`thiserror`](https://github.com/dtolnay/thiserror) would make sense for this? 

Using those crates would save you from writing boilerplate code and keeping all that boilerplate code up to date when adding new error sources.

---

Also a more general note: I saw that you have a `main.rs` and then a `gccrs` module. Usually you just make a lib and add one small `main.rs` to keep this split. You could do this, by moving everything from `src/gccrs/*` to `src/` and rename `mod.rs` to `lib.rs`

**Comment by CohenArthur** [comment]
> Did you look into if using [`anyhow`](https://github.com/dtolnay/anyhow) and/or [`thiserror`](https://github.com/dtolnay/thiserror) would make sense for this?
> 
> Using those crates would save you from writing boilerplate code and keeping all that boilerplate code up to date when adding new error sources.

I didn't look into it and have never used them yet, but I'm open to trying! I think `thiserror` would be simpler to work with and contribute to.

> Also a more general note: I saw that you have a `main.rs` and then a `gccrs` module. Usually you just make a lib and add one small `main.rs` to keep this split. You could do this, by moving everything from `src/gccrs/*` to `src/` and rename `mod.rs` to `lib.rs`

Fair point. I didn't know at the time if we were gonna need multiple modules or not. I think it's established now that it won't be necessary, so I'll do this. Maybe in a future PR (I'll open an issue) as to not clutter this one

**Comment by flip1995** [comment]
You usually want to use `thiserror` for the library part and then `anyhow` in the binary part. I think this is a good idea, since you'll get good looking error messages for free.

**Review by philberty — APPROVED** [review]
LGTM

**Comment by CohenArthur** [comment]
I cleaned up the error handling but this is a temporary and ugly solution. This way, as @flip1995 said, we get good looking error messages and handling for free. Once the project is split into binary and library (#22), this will get refactored. There will be a single error type (not a `WrapperError` and a `gccrs::Error` one), and the binary will not need to map errors etc as it does now.

**Comment by flip1995** [comment]
How much work do you think it would be to do the split between lib and binary before merging this PR? AFAICT this shouldn't be that hard to do. I'd rather do that first and then do this PR right once, instead of merging this PR -> do the split -> merging a PR refactoring this PR.

**Comment by CohenArthur** [comment]
> How much work do you think it would be to do the split between lib and binary before merging this PR? AFAICT this shouldn't be that hard to do. I'd rather do that first and then do this PR right once, instead of merging this PR -> do the split -> merging a PR refactoring this PR.

You make a good point haha. I'll do the split here and rename the PR then

File: src/lib.rs
Comment (flip1995) [inline src/lib.rs]: Why do you need a wrapper for this? Do you plan to do more things with this wrapper in the future? If not, I'd just implement the `spawn()` function here directly.

File: src/lib.rs
Comment (flip1995) [inline src/lib.rs]: I'd rather name this `gccrs-driver` to stick to the rustc naming.

File: src/main.rs
Comment (flip1995) [inline src/main.rs]: Wait, does this handle both the `driver` and the `cargo` part? So if you call `cargo gccrs build`, this binary will execute itself with the same arguments?

When is this called with `gccrs` vs `rustc`?

**Review by flip1995 — CHANGES_REQUESTED** [review]
I figured that this is now the right time to do a whole project review, and not only a review of this PR. My suggestions don't have to be implemented all in one/this PR, but rather get implemented step by step. I tried to sort them from easy to hard to do.

---

What is this line supposed to do? https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/gccrs.rs#L23

---

https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/gccrs.rs#L43-L44

You can use the [`which`](https://crates.io/crates/which) crate for cross-platform support for this 

---

I think you want to have 2 binaries here. One for the driver, that does all the argument conversion and other hard work, and one that just invokes the driver from the cargo command. But I may be misunderstanding something, see my comment https://github.com/Rust-GCC/cargo-gccrs/pull/21#discussion_r665144475

---

https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/config.rs#L1-L4

What format is the config dump in and can you use serde for this?

---

https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/config.rs#L101-L102

A fallible display impl seems weird to me. You should rather implement a fallible constructor and then the `Display` trait. That way you can use the format machinery and get functions like `to_string` for free. 

I'm also not 100% sure what the option dumping is for, so please tell me if there is a reason why it has to be done that way.

---

https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/args.rs#L171

I still don't quite get why it is necessary to return a `Vec<GccrsArgs>`. Just from the struct name and the documentation, I would assume that when I have a `GccrsArgs` instance, that I have all the arguments that I need. This is also the main thing that prevents the `GccrsArgs` from implementing things like the `TryFrom` trait and so on. 

---

https://github.com/Rust-GCC/cargo-gccrs/blob/e47fe3b6badb81f377efe2138dd0fbcff79ec36c/src/args.rs#L164-L168

I still think that with this, `GccrsArgs` is doing too much. One good principle of OOP is that one class should only do one thing. With that principle, `GccrsArgs` should only handle arguments, which includes parsing, converting, and then holding them. Now it handles arguments and does random other things like creating archives and in the future possibly even more.

This is the area of the current implementation that really needs a refactor IMHO. Apply the KISS principle as often as you can. I have trouble following what is done where, so the current impl isn't really simple to follow.

**Comment by flip1995** [comment]
Let's track and further talk about this in #36 and revisit this PR once the other points are addressed. Fixing everything in #36 should bring this to a point, where this split can just be done by moving files around.

File: src/lib.rs
Comment (CohenArthur) [reply src/lib.rs]: That's what I had done originally but thought it'd be cleaner from the binary side of things. I'm fine with reverting to a simple function named `spawn_as_wrapper()` as it was originally

File: src/main.rs
Comment (bjorn3) [reply src/main.rs]: When you do `cargo gccrs`, cargo will run `cargo-gccrs gccrs`. When you do `RUSTC_WRAPPER=cargo-gccrs`, the spawned rustc processes will be `cargo-gccrs rustc`. (`RUSTC` instead of `RUSTC_WRAPPER` would omit the `rustc` part)

**Comment by CohenArthur** [comment]
@flip1995 I believe that all items have been resolved. I also took the liberty of splitting the wrapper in two binaries as mentioned in your big review comment, since it ties in nicely with the PR.

**Review by philberty — APPROVED** [review]
LGTM good work

File: src/bin-driver.rs
Comment (flip1995) [inline src/bin-driver.rs:5]: Some module level documentation may be good here to explain the difference between the driver and the wrapper.

File: src/bin-driver.rs
Comment (flip1995) [inline src/bin-driver.rs]: A function call named `handle_rust_args` that also compiles things is a bit confusing IMO.

**Review by flip1995 — APPROVED** [review]
LGTM overall. My 2 comments shouldn't block this PR from going through.

File: src/bin-driver.rs
Comment (CohenArthur) [reply src/bin-driver.rs]: That's fair. Do you have any idea as to what to call the function? Since it doesn't only compile, and also has the task of translating the Rust arguments

Comment (flip1995) [reply src/bin-driver.rs]: I would either split it up and first handle the args and then compile or just call it ``compile`` if you want to keep only one public function to interface with the lib part.

File: src/gccrs.rs
Comment (flip1995) [inline src/gccrs.rs]: Now the method name sounds like it compiles _the arguments_. Why not just "`compile(args)`"?

Comment (CohenArthur) [reply src/gccrs.rs]: I have a bit of an issue with `compile` because it sounds like we're simply using some sort of gccrs wrapper, which would logically take gccrs arguments only. I would be okay with `compile` since the gccrs module will not become its own crate or anything, but I feel weird looking at `Gccrs::compile(args)` when the args in question are clearly `rustc` args and not `gcc` args.

I don't know, I thought this was a good compromise but I'm okay with renaming it as just `compile` if you'd like

Comment (flip1995) [reply src/gccrs.rs]: That's a good point. What about the name `compile_with_rust_args`? So adding a "with". (This is really NIT picking, so if you don't want to change it, but just merge it, I'd also be ok with that)

Comment (flip1995) [reply src/gccrs.rs]: That's a good point. What about the name `compile_with_rust_args`? So adding a "with". (This is really NIT picking, so if you don't want to change it, but just merge it, I'd also be ok with that)

Comment (CohenArthur) [reply src/gccrs.rs]: Perfect! This is what I was thinking too this morning :)

**Review by flip1995 — APPROVED** [review]
Great, let's get this merged then!

**Pull request merged**
